### PR TITLE
[6.0] Fix for downcasting BigInt to Long fails

### LINKED ISSUE #877: Downcasting BigInt to Long fails
When downcasting a BigInt to Long it fails with a runtime exception, no matter the value.

Example ergoscript that would trigger the error:
```
{
  sigmaProp(1L.toBigInt.toLong < CONTEXT.preHeader.timestamp)
}
```
Resulting error:
java.lang.RuntimeException: Cannot downcast value CBigInt(1) to the type SLong$

### PR DESCRIPTION
**Pull request by kushti**
Close #877

It has test from the issue as well similar ones for all the Numeric types.

Also, Downcast from BigInt to BigInt and Upcast from BigInt to BigInt are fixed now

### CODE REVIEW

File: sc/shared/src/test/scala/sigmastate/TestingInterpreterSpecification.scala [hunk 1]
```diff
@@ -215,6 +216,46 @@ class TestingInterpreterSpecification extends CompilerTestingCommons
...
+      test()
+    } else {
+      an[Exception] shouldBe thrownBy(test())
+    }
+  }
+
+  property("upcasting to bigint") {
+    testEval("{ sigmaProp(1L.toBigInt < bigInt(\"2\")) }")
+  }
+
```
Comment (aslesarenko) [inline sc/shared/src/test/scala/sigmastate/TestingInterpreterSpecification.scala:258]: Every v6.0 change should be tested in LanguageSpecificationV6, this is the only way to ensure:
1) existing behavior is preserved for old versions (including required failures)
2) the new behavior is enabled starting from the given version
2) the implementation is complete, i.e. all the compiler and interpreter components work as expected
In this specific case, changeFeature can be used.

Separate unit tests cannot ensure test coverage.

**Review by aslesarenko — CHANGES_REQUESTED** [review]

File: sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala [hunk 2]
```diff
@@ -172,21 +172,24 @@ class LanguageSpecificationV6 extends LanguageSpecificationBase { suite =>
...
-    )
+      // The `Downcast(bigInt, SBigInt)` node is never produced by ErgoScript compiler, but is still valid ErgoTree.
+      // Fixed in 6.0
+      assertExceptionThrown(
+        SBigInt.downcast(CBigInt(new BigInteger("0", 16)).asInstanceOf[AnyVal]),
+        _.getMessage.contains("Cannot downcast value")
+      )
+    } else {
+      SBigInt.upcast(CBigInt(new BigInteger("0", 16)).asInstanceOf[AnyVal]) shouldBe CBigInt(new BigInteger("0"))
+      SBigInt.downcast(CBigInt(new BigInteger("0", 16)).asInstanceOf[AnyVal]) shouldBe CBigInt(new BigInteger("0"))
```
[the remark this answers is not in the document]
Comment (kushti) [reply sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala]: done

File: sc/shared/src/test/scala/sigmastate/TestingInterpreterSpecification.scala (hunk 1 above)
Comment (kushti) [reply sc/shared/src/test/scala/sigmastate/TestingInterpreterSpecification.scala:258]: this test is non-versioned, so nothing to bring to LSV6 from it

For versioned tests, alternatives are added to LSV6.

**Comment by kushti** [comment]
@aslesarenko comments addressed , please make another pass

File: sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala [hunk 3]
```diff
@@ -172,21 +172,26 @@ class LanguageSpecificationV6 extends LanguageSpecificationBase { suite =>
...
-    assertExceptionThrown(
-      SBigInt.upcast(CBigInt(new BigInteger("0", 16)).asInstanceOf[AnyVal]),
-      _.getMessage.contains("Cannot upcast value")
-    )
+    if (activatedVersionInTests < VersionContext.V6SoftForkVersion) {
+      // The `Upcast(bigInt, SBigInt)` node is never produced by ErgoScript compiler, but is still valid ErgoTree.
+      // Fixed in 6.0
+      assertExceptionThrown(
+        SBigInt.upcast(CBigInt(new BigInteger("0", 16)).asInstanceOf[AnyVal]),
+        _.getMessage.contains("Cannot upcast value")
```
Comment (aslesarenko) [inline sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala:180]: good enough, but FYI there is also method `exceptionLike[ConcreteExectionType]("Cannot upcast value")` which tests both exception type and the contents of the message.

**Review by aslesarenko — APPROVED** [review]
LGTM, but please see the additional comment.

File: sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala (hunk 3 above)
Comment (kushti) [reply sc/shared/src/test/scala/sigma/LanguageSpecificationV6.scala:180]: The check is corresponding to previously written code, so such refactoring is a subject of a dedicated issue & PR I suppose

**Pull request merged**